File: mobile/lib/features/games/domain/config/emotional_radar_v2_config.dart
/// Miroir Dart de `EmotionalRadarV2Config` (backend) — « Emotional Radar v2 ».
///
/// Toute modification ici impose la même modification côté backend (AGENTS.md §7.7).
/// Table de configuration admin du brief (2026-08-12, niveaux de choix 6/6/9/9).
library;

/// Bande de proximité sémantique visée par un niveau (valeurs PROVISOIRES).
enum DistanceBand {
  high(0.70), // émotions très différentes (facile)
  medium(0.45),
  low(0.22); // émotions très proches (difficile)

  const DistanceBand(this.target);
  final double target;
}

/// Un niveau à deux axes : nombre de choix (charge) + distance visée (finesse).
class DifficultyLevel {
  const DifficultyLevel(this.level, this.choicesCount, this.targetDistance);
  final int level;
  final int choicesCount;
  final DistanceBand targetDistance;
}

class EmotionalRadarV2Config {
  EmotionalRadarV2Config._();

  static const int emotionPoolSize = 45;
  static const int totalScenes = 15;
  static const int videoLibrarySize = 135; // 45 × 3 intensités
  static const List<String> stimulusIntensityLevels = ['Faible', 'Modérée', 'Intense'];
  static const List<String> intensityScale = stimulusIntensityLevels;

  static const int difficultyLevels = 4;

  /// 6 / 6 / 9 / 9 + distance Élevée / Moyenne / Élevée / Faible.
  static const List<DifficultyLevel> levels = [
    DifficultyLevel(1, 6, DistanceBand.high),
    DifficultyLevel(2, 6, DistanceBand.medium),
    DifficultyLevel(3, 9, DistanceBand.high), // charge cognitive isolée
    DifficultyLevel(4, 9, DistanceBand.low), // charge + finesse combinées
  ];

  static const double levelUpThreshold = 0.70;
  static const double levelDownThreshold = 0.40;
  static const int evaluationWindowMin = 3;
  static const int evaluationWindowMax = 4;
  static const int startingLevel = 1;

  /// Budget de réponse d'une scène.
  ///
  /// Le référentiel écrit 8 000 ms ; porté à 30 000 ms à la demande du client.
  /// Ce budget couvre le visionnage PUIS la réponse, et les clips existants
  /// vont jusqu'à 12 s — 8 s ne laissaient même pas voir la scène en entier.
  static const int maxResponseTimeMs = 30000;
  static const int minImpulsiveTimeMs = 400;

  static const bool normingRequiredBeforeUse = true;
  /// Justification écrite demandée au joueur.
  ///
  /// Passé à `false` : le client a retiré la troisième question de l'écran. Le
  /// référentiel la demandait après chaque réponse — l'écart est assumé, et
  /// `justificationScoringAvailable` reste faux côté serveur tant qu'aucun
  /// texte n'est collecté, de sorte que le rapport ne prétende pas la noter.
  static const bool requireExplanation = false;
  static const int justificationMinScore = 0;
  static const int justificationMaxScore = 5;

  static DifficultyLevel level(int levelNumber) => levels[levelNumber - 1];
  static int choicesForLevel(int levelNumber) => level(levelNumber).choicesCount;
}

File: backend/src/main/java/com/zennyt/games/domain/config/EmotionalRadarV2Config.java
package com.zennyt.games.domain.config;

import com.zennyt.games.domain.vo.DifficultyLevel;
import com.zennyt.games.domain.vo.DistanceBand;

import java.util.List;

/**
 * Configuration <b>MOTEUR</b> d'« Emotional Radar v2 » — reprend la table de
 * configuration admin du brief (2026-08-12, « niveaux de choix 6/6/9/9 »). Java pur.
 *
 * <p>Deux axes de difficulté indépendants (charge cognitive = nombre de choix ;
 * finesse = proximité sémantique). Difficulté <b>adaptative</b> : montée &gt; 70 %,
 * descente &lt; 40 %, sur une fenêtre glissante de 3–4 scènes. Les valeurs de bande
 * de distance ({@link DistanceBand}) et le barème du score « jeu » vivent dans la
 * couche provisoire ; ici on ne code que la structure fournie par le brief.
 */
public final class EmotionalRadarV2Config {

    private EmotionalRadarV2Config() {
    }

    // ── Référentiel émotionnel (admin table) ─────────────────────────────────

    /** {@code emotion_pool_size} — 45 émotions (Cowen &amp; Keltner + littérature). */
    public static final int EMOTION_POOL_SIZE = 45;

    /** {@code total_scenes} — scènes par session. */
    public static final int TOTAL_SCENES = 15;

    /** {@code video_library_size} — 45 émotions × 3 intensités. */
    public static final int VIDEO_LIBRARY_SIZE = 135;

    /** Niveaux d'intensité réellement joués dans la vidéo. */
    public static final List<String> STIMULUS_INTENSITY_LEVELS =
        List.of("Faible", "Modérée", "Intense");

    /** Échelle d'intensité perçue par le joueur (distincte de l'intensité produite). */
    public static final List<String> INTENSITY_SCALE = STIMULUS_INTENSITY_LEVELS;

    // ── Difficulté adaptative (admin table) ──────────────────────────────────

    /** {@code difficulty_levels} — 4 niveaux. */
    public static final int DIFFICULTY_LEVELS = 4;

    /**
     * {@code choices_per_level} — 6 / 6 / 9 / 9 (mis à jour depuis 6/6/8/8, grille 3×3).
     * Combiné à {@code target_distance_per_level} = Élevée / Moyenne / Élevée / Faible.
     */
    public static final List<DifficultyLevel> LEVELS = List.of(
        new DifficultyLevel(1, 6, DistanceBand.HIGH),    // facile
        new DifficultyLevel(2, 6, DistanceBand.MEDIUM),  // finesse moyenne
        new DifficultyLevel(3, 9, DistanceBand.HIGH),    // charge cognitive isolée
        new DifficultyLevel(4, 9, DistanceBand.LOW)      // charge + finesse combinées
    );

    /** {@code level_up_threshold_percent} — réussite pour monter de niveau. */
    public static final double LEVEL_UP_THRESHOLD = 0.70;

    /** {@code level_down_threshold_percent} — réussite en dessous de laquelle on redescend. */
    public static final double LEVEL_DOWN_THRESHOLD = 0.40;

    /** {@code evaluation_window_scenes} — fenêtre glissante (min..max). */
    public static final int EVALUATION_WINDOW_MIN = 3;
    public static final int EVALUATION_WINDOW_MAX = 4;

    /** Niveau de départ par défaut d'une session. */
    public static final int STARTING_LEVEL = 1;

    // ── Timing (admin table) ─────────────────────────────────────────────────

    /** {@code max_response_time_ms} — temps max autorisé pour répondre. */
    public static final int MAX_RESPONSE_TIME_MS = 8000;

    /** {@code min_impulsive_time_ms} — réponse trop rapide (absence d'analyse). */
    public static final int MIN_IMPULSIVE_TIME_MS = 400;

    // ── Norming / gouvernance (admin table) ──────────────────────────────────

    /** {@code norming_required_before_use} — chaque vidéo validée par un panel. */
    public static final boolean NORMING_REQUIRED_BEFORE_USE = true;

    /** {@code require_explanation} — justification textuelle demandée au joueur (0–5). */
    public static final boolean REQUIRE_EXPLANATION = true;

    public static final int JUSTIFICATION_MIN_SCORE = 0;
    public static final int JUSTIFICATION_MAX_SCORE = 5;

    /** Niveau par son numéro (1..4). */
    public static DifficultyLevel level(int levelNumber) {
        if (levelNumber < 1 || levelNumber > LEVELS.size()) {
            throw new IllegalArgumentException("niveau hors 1.." + LEVELS.size() + " : " + levelNumber);
        }
        return LEVELS.get(levelNumber - 1);
    }

    /** Nombre de choix proposés au niveau donné. */
    public static int choicesForLevel(int levelNumber) {
        return level(levelNumber).choicesCount();
    }
}
